packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -0,0 +1,191 @@
+import { PendingStateManager } from "../pendingStateManager";
+import { BatchManager } from "./batchManager";
+import { IOutboxOptions, IBatchProcessors, BatchMessage, IBatch } from "./definitions";
+
+export class Outbox {
+    private readonly attachFlowBatch: BatchManager;
+    private readonly mainBatch: BatchManager;
+    private readonly defaultAttachFlowSoftLimitInBytes = 64 * 1024;
+
+    constructor(
+        private readonly shouldSend: () => boolean,
+        private readonly pendingStateManager: PendingStateManager,
+        private readonly context: IContainerContext,
+        private readonly options: IOutboxOptions,
+        private readonly batchProcessors: IBatchProcessors,
+    ) {
+        // We need to allow infinite size batches if we enable compression
+        const hardLimit = this.options.compressionOptions?.minimumBatchSizeInBytes !== Infinity
+            ? Infinity : this.options.maxBatchSizeInBytes;
+        const softLimit = this.options.compressionOptions?.minimumBatchSizeInBytes !== Infinity
+            ? Infinity : this.defaultAttachFlowSoftLimitInBytes;
+
+        this.attachFlowBatch = new BatchManager({
+            hardLimit,
+            softLimit,
+        });
+        this.mainBatch = new BatchManager({
+            hardLimit
+        });
+    }
+
+    public get isEmpty(): boolean {
+        return this.attachFlowBatch.length === 0 && this.mainBatch.length === 0;
+    }
+
+    public submit(message: BatchMessage) {
+        if (!this.mainBatch.push(message)) {
+            throw new GenericError(
+                "BatchTooLarge",
+                /* error */ undefined,
+                {
+                    opSize: (message.contents?.length) ?? 0,
+                    count: this.mainBatch.length,
+                    limit: this.mainBatch.options.hardLimit,
+                });
+        }
+    }
+
+    public submitAttach(message: BatchMessage) {
+        if (!this.attachFlowBatch.push(message)) {
+            // BatchManager has two limits - soft limit & hard limit. Soft limit is only engaged
+            // when queue is not empty.
+            // Flush queue & retry. Failure on retry would mean - single message is bigger than hard limit
+            this.flushInternal(this.attachFlowBatch.popBatch());
+            if (!this.attachFlowBatch.push(message)) {
+                throw new GenericError(
+                    "BatchTooLarge",
+                    /* error */ undefined,
+                    {
+                        opSize: (message.contents?.length) ?? 0,
+                        count: this.attachFlowBatch.length,
+                        limit: this.attachFlowBatch.options.hardLimit,
+                    });
+            }
+        }
+    }
+
+    public flush() {
+        this.flushInternal(this.attachFlowBatch.popBatch());
+        this.flushInternal(this.mainBatch.popBatch());
+    }
+
+    private flushInternal(batch: IBatch) {
+        this.flushBatch(
```

Comment:
It's somewhat weird to see flushInternal() call flushBatch() - they are both internal (private) and both flush batch (in some extend). Can you find better naming that more clearly articulates what they are doing (such that it's more obvious without looking at implementation who should be calling whom)?